src/server/server.hpp

```diff
@@ -17,42 +17,70 @@
 
 #include "config.hpp"
 #include "user_handler.hpp"
-#include "input_stream.hpp"
 #include "metrics.hpp"
-#include "network.hpp"
 #include "ban.hpp"
 #include "player.hpp"
-#include "room_manager.hpp"
 #include "simple_wml.hpp"
+#include "player_connection.hpp"
 
 #include <boost/scoped_ptr.hpp>
+#include <boost/shared_array.hpp>
 #include <boost/ptr_container/ptr_vector.hpp>
+
+#include <boost/asio/signal_set.hpp>
+
+namespace wesnothd
+{
+
 class server
 {
 public:
-	server(int port, const std::string& config_file, size_t min_threads,size_t max_threads);
+	server(int port, bool keep_alive, const std::string& config_file, size_t min_threads,size_t max_threads);
 	void run();
 private:
-	void send_error(network::connection sock, const char* msg, const char* error_code ="") const;
-	void send_error(network::connection sock, const std::string &msg, const char* error_code = "") const
-	{
-		send_error(sock, msg.c_str(), error_code);
-	}
-
-	void send_warning(network::connection sock, const char* msg, const char* warning_code ="") const;
-	void send_warning(network::connection sock, const std::string &msg, const char* warning_code = "") const
-	{
-		send_warning(sock, msg.c_str(), warning_code);
+	boost::asio::io_service io_service_;
+	boost::asio::ip::tcp::acceptor acceptor_;
+	void serve();
+	void accept_connection(const boost::system::error_code& error, socket_ptr socket);
+
+	union {
+		boost::uint32_t connection_num;
+		char buf[4];
+	} handshake_response_;
```

Comment:
It also should directly send/receive gzipped wml sizes but I took care to feed those to htonl()/ntohl()